File: core/data/src/main/java/me/nalab/core/data/survey/Survey.java
package me.nalab.core.data.survey;

import java.util.List;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.OneToMany;
import javax.persistence.Table;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;
import me.nalab.core.data.common.TimeBaseEntity;

@Entity
@Table(name = "survey")
@Getter
@Setter
@SuperBuilder
@NoArgsConstructor
@AllArgsConstructor
public class Survey extends TimeBaseEntity {

	@Id
	@Column(name = "survey_id")
	private Long id;

	@OneToMany(mappedBy = "survey", fetch = FetchType.LAZY)
	private List<FormQuestion> formQuestionableList;

	@JoinColumn(name = "target_id", nullable = false)
	private Long targetId;

}

File: survey/src/main/kotlin/me/nalab/survey/domain/survey/Survey.kt
package me.nalab.survey.domain.survey

import me.nalab.core.data.common.TimeBaseEntity
import me.nalab.survey.domain.survey.exception.DuplicatedOrderException
import javax.persistence.*

@Table(name = "survey")
@Entity(name = "survey")
class Survey(
    @Id
    @Column(name = "gallery_id")
    val id: Long,

    @OneToMany(
        mappedBy = "survey",
        fetch = FetchType.LAZY,
        cascade = [CascadeType.PERSIST, CascadeType.MERGE]
    )
    val formQuestionables: MutableList<FormQuestionable>,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "target_id")
    val target: Target,
) : TimeBaseEntity() {

    init {
        formQuestionables.sorted()
        validNoDuplicatedFormOrder()
    }

    private fun validNoDuplicatedFormOrder() {
        val orders = HashSet<Int>()
        this.formQuestionables.forEach { formQuestion ->
            val order: Int = formQuestion.order
            if (orders.contains(order)) {
                throw DuplicatedOrderException(order, orders)
            }
            orders.add(order)
        }
    }
}